File: typescript/sdk/examples/attribution-gateway/src/fixture-backend.ts
#!/usr/bin/env node
// Copyright 2025 The MCP Interceptors Authors. All rights reserved.
/**
 * Fixture skill backend — a tiny self-contained stdio MCP server that serves a
 * few `skill://…/SKILL.md` resources from inline fixtures. Stands in for a real
 * skills server (like fallout-helper) so the attribution-gateway example runs
 * anywhere with no external repo or build. Three skills exercise each grade:
 * compliant (layered attribution), partial (author only), and uncredited.
 */
import { Server } from '@modelcontextprotocol/sdk/server/index.js';
import { StdioServerTransport } from '@modelcontextprotocol/sdk/server/stdio.js';
import {
  ListResourcesRequestSchema,
  ReadResourceRequestSchema,
} from '@modelcontextprotocol/sdk/types.js';

const COMPLIANT_SKILL = `---
name: weather-reporter
description: Look up and summarize current weather for a location.
license: Apache-2.0
metadata:
  skill_author:
    name: Jane Dev
    url: https://github.com/janedev
  version: 1.2.0
  sources:
    - title: OpenWeather API documentation
      publisher: OpenWeather
      url: https://openweathermap.org/api
      rights_basis: license_grant
      covers: weather field semantics and units
  attribution: |
    Authored by Jane Dev (Apache-2.0). Weather field semantics derived from the
    OpenWeather API documentation under their developer terms.
---
# Weather Reporter
Body content...
`;

const PARTIAL_SKILL = `---
name: note-taker
description: Capture and organize freeform notes.
metadata:
  skill_author: Sam Maker
---
# Note Taker
Body content...
`;

const UNCREDITED_SKILL = `---
name: mystery-tool
description: Does something useful; provenance unclear.
---
# Mystery Tool
Body content...
`;

interface SkillFixture {
  name: string;
  description: string;
  text: string;
}

const SKILLS: Record<string, SkillFixture> = {
  'skill://demo/weather-reporter/SKILL.md': {
    name: 'weather-reporter',
    description: 'Weather lookup skill (fully attributed).',
    text: COMPLIANT_SKILL,
  },
  'skill://demo/note-taker/SKILL.md': {
    name: 'note-taker',
    description: 'Note-taking skill (author only).',
    text: PARTIAL_SKILL,
  },
  'skill://demo/mystery-tool/SKILL.md': {
    name: 'mystery-tool',
    description: 'Unattributed skill.',
    text: UNCREDITED_SKILL,
  },
};

const server = new Server(
  { name: 'fixture-skill-backend', version: '1.0.0' },
  { capabilities: { resources: {} } },
);

server.setRequestHandler(ListResourcesRequestSchema, async () => ({
  resources: Object.entries(SKILLS).map(([uri, s]) => ({
    uri,
    name: s.name,
    description: s.description,
    mimeType: 'text/markdown',
  })),
}));

server.setRequestHandler(ReadResourceRequestSchema, async (req) => {
  const uri = req.params.uri;
  const skill = SKILLS[uri];
  if (!skill) throw new Error(`Unknown resource: ${uri}`);
  return { contents: [{ uri, mimeType: 'text/markdown', text: skill.text }] };
});

await server.connect(new StdioServerTransport());

File: rust/demos/skill-attribution-mcp/src/fixture_backend.rs
// Copyright 2025 The MCP Interceptors Authors. All rights reserved.
// Use of this source code is governed by an Apache-2.0
// license that can be found in the LICENSE file.

//! A self-contained `rmcp` MCP server that serves three inline `SKILL.md`
//! resources — compliant, partial, and uncredited — so the demo has a real MCP
//! backend with no external dependencies. The skill bodies match the SDK's
//! `attribution_fixtures` example, so the grades come out identical.

use rmcp::{
    model::*, service::RequestContext, service::RunningService, ErrorData as McpError, RoleClient,
    RoleServer, ServerHandler, ServiceExt,
};
use serde_json::json;

/// A compliant skill: author + license + a layered `sources[]` chain backed by
/// an `attribution` block → `compliant_with_upstream_attribution`.
const COMPLIANT: &str = r#"---
name: fallout-rpg
description: Fallout tabletop RPG helper
license: CC-BY-4.0
metadata:
  skill_author: Vault-Tec Archivists
  version: 1.2.0
  sources:
    - https://fallout.bethesda.net
    - https://falloutwiki.com
  attribution: "Derived from the Fallout tabletop community SRD."
  citations:
    - Fallout RPG Core Rulebook
---
# Fallout RPG
Roll for initiative, vault dweller.
"#;

/// A partially-attributed skill: author only, no license or provenance →
/// `partial` (with a `warn` on the missing license).
const PARTIAL: &str = r#"---
name: note-taker
description: A simple note taker
metadata:
  skill_author: J. Doe
---
# Note Taker
Jot it down.
"#;

/// An uncredited skill: no YAML frontmatter at all → `non-compliant`, and an
/// `error` that blocks the chain (SEP-2640 conformance failure).
const UNCREDITED: &str = r#"# Mystery Tool
Does something, attributed to no one.
"#;

/// The skills this backend serves: `(name, SKILL.md body)`.
const SKILLS: &[(&str, &str)] = &[
    ("fallout-rpg", COMPLIANT),
    ("note-taker", PARTIAL),
    ("mystery-tool", UNCREDITED),
];

fn uri_for(name: &str) -> String {
    format!("skill://demo/{name}/SKILL.md")
}

/// The fixture MCP server.
#[derive(Clone)]
pub struct FixtureBackend;

impl ServerHandler for FixtureBackend {
    fn get_info(&self) -> ServerInfo {
        ServerInfo::new(ServerCapabilities::builder().enable_resources().build()).with_instructions(
            "Fixture skill backend serving three SKILL.md manifests.".to_string(),
        )
    }

    async fn list_resources(
        &self,
        _request: Option<PaginatedRequestParams>,
        _: RequestContext<RoleServer>,
    ) -> Result<ListResourcesResult, McpError> {
        let resources = SKILLS
            .iter()
            .map(|(name, _)| {
                RawResource::new(uri_for(name), format!("{name} skill manifest")).no_annotation()
            })
            .collect();
        Ok(ListResourcesResult {
            resources,
            next_cursor: None,
            meta: None,
        })
    }

    async fn read_resource(
        &self,
        request: ReadResourceRequestParams,
        _: RequestContext<RoleServer>,
    ) -> Result<ReadResourceResult, McpError> {
        for (name, text) in SKILLS {
            if request.uri == uri_for(name) {
                return Ok(ReadResourceResult::new(vec![ResourceContents::text(
                    *text,
                    request.uri.clone(),
                )]));
            }
        }
        Err(McpError::resource_not_found(
            "resource_not_found",
            Some(json!({ "uri": request.uri })),
        ))
    }
}

/// Spin up the fixture backend on an in-memory transport and return a connected
/// MCP client wired to it — no child process, no stdio (robust on Windows).
pub async fn connect_fixture() -> Result<RunningService<RoleClient, ()>, Box<dyn std::error::Error>>
{
    let (server_transport, client_transport) = tokio::io::duplex(8192);

    tokio::spawn(async move {
        if let Ok(server) = FixtureBackend.serve(server_transport).await {
            let _ = server.waiting().await;
        }
    });

    let client = ().serve(client_transport).await?;
    Ok(client)
}
